File: example/kalman/kalman_filter.cpp
#include "kalman_filter.h"

KalmanFilter::KalmanFilter(
    const Matrix<double, 8, 1>& initialState,
    const Matrix<double, 8, 8>& initialCovariance,
    const Matrix<double, 8, 8>& initialTransition,
    const Matrix<double, 6, 8>& initialMeasurement,
    const Matrix<double, 6, 6>& initialMeasurementNoiseCov,
    const Matrix<double, 8, 8>& initialProcessNoiseCov)
    : state_(initialState),
      covariance_(initialCovariance),
      transitionMatrix_(initialTransition),
      measurementMatrix_(initialMeasurement),
      measurementNoiseCov_(initialMeasurementNoiseCov),
      processNoiseCov_(initialProcessNoiseCov) {}

Matrix<double, 8, 1> KalmanFilter::getState() const { return state_; }

void KalmanFilter::predict(const Matrix<double, 6, 1>& measurement,
                           const double dt) {
  transitionMatrix_(0, 1) = dt;
  transitionMatrix_(1, 2) = dt;
  transitionMatrix_(3, 4) = dt;
  transitionMatrix_(4, 5) = dt;
  transitionMatrix_(6, 7) = dt;
  state_ = transitionMatrix_ * state_;
  covariance_ = transitionMatrix_ * covariance_ * transitionMatrix_.t() +
                processNoiseCov_;
  Matrix<double, 6, 1> innovation = measurement - measurementMatrix_ * state_;
  Matrix<double, 6, 6> innovationCov =
      measurementMatrix_ * covariance_ * measurementMatrix_.t() +
      measurementNoiseCov_;
  Matrix<double, 8, 6> kalmanGain =
      covariance_ * measurementMatrix_.t() * innovationCov.inv();
  state_ = state_ + kalmanGain * innovation;
  covariance_ = covariance_ - kalmanGain * measurementMatrix_ * covariance_;
}

int main() {
  // Define initial matrices for Kalman filter
  Matrix<double, 8, 1> initialState{
      0, 12, 0.1, 0, 0.2, 0.1, 3.14, 0.12,
  };
  Eye<double, 8> initialCovariance;
  Eye<double, 8> initialTransition;
  Matrix<double, 6, 8> initialMeasurement;
  initialMeasurement(0, 1) = 1;
  initialMeasurement(1, 2) = 1;
  initialMeasurement(2, 4) = 1;
  initialMeasurement(3, 5) = 1;
  initialMeasurement(4, 6) = 1;
  initialMeasurement(5, 7) = 1;
  Eye<double, 6> initialMeasurementNoiseCov;
  initialMeasurementNoiseCov *= 0.1;
  Eye<double, 8> initialProcessNoiseCov;
  initialProcessNoiseCov *= 0.01;

  // Initialize the Kalman filter
  KalmanFilter kf(initialState, initialCovariance, initialTransition,
                  initialMeasurement, initialMeasurementNoiseCov,
                  initialProcessNoiseCov);

  // Perform Kalman filter operations
  Matrix<double, 6, 1> measurement{
      12.0, 0.2, 0.1, 0.1, 3.141, 0.11,
  };
  double dt = 0.15;

  kf.predict(measurement, dt);

  std::cout << kf.getState() << std::endl;

  return 0;
}

File: example/kalman/kalman_filter.py
import numpy as np


def kalman_filter(x, P, F, H, R, Q, measurement, dt):
    # Update state transition matrix based on dynamic time step
    F[0, 1] = dt
    F[1, 2] = dt
    F[3, 4] = dt
    F[4, 5] = dt
    F[6, 7] = dt

    # Prediction step
    x_predicted = np.dot(F, x)
    P_predicted = np.dot(np.dot(F, P), F.T) + Q

    # Update step
    y = measurement - np.dot(H, x_predicted)
    S = np.dot(np.dot(H, P_predicted), H.T) + R
    K = np.dot(np.dot(P_predicted, H.T), np.linalg.inv(S))
    x = x_predicted + np.dot(K, y)
    P = np.dot((np.eye(x.shape[0]) - np.dot(K, H)), P_predicted)
    return x, P


# Initial error covariance matrix
P = np.eye(8)

# Define the state transition matrix
F = np.array(
    [
        [1, 1, 0, 0, 0, 0, 0, 0],
        [0, 1, 1, 0, 0, 0, 0, 0],
        [0, 0, 1, 0, 0, 0, 0, 0],
        [0, 0, 0, 1, 1, 0, 0, 0],
        [0, 0, 0, 0, 1, 1, 0, 0],
        [0, 0, 0, 0, 0, 1, 0, 0],
        [0, 0, 0, 0, 0, 0, 1, 1],
        [0, 0, 0, 0, 0, 0, 0, 1],
    ],
    dtype=float,
)

# Define the measurement matrix (speed, acceleration, yaw, yaw rate)
H = np.array(
    [
        [0, 1, 0, 0, 0, 0, 0, 0],
        [0, 0, 1, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 1, 0, 0, 0],
        [0, 0, 0, 0, 0, 1, 0, 0],
        [0, 0, 0, 0, 0, 0, 1, 0],
        [0, 0, 0, 0, 0, 0, 0, 1],
    ]
)

# Define measurement noise covariance
R = np.eye(6) * 0.1

# Define process noise covariance
Q = np.eye(8) * 0.01

kalman_state = np.array(
    [
        0,
        12,
        0.1,
        0,
        0.2,
        0.1,
        3.14,
        0.12,
    ]
).T  # Initial x, vx, ax, y, vy, ay, yaw, yaw rate

measurement = np.array(
    [
        12.0,
        0.2,
        0.1,
        0.1,
        3.141,
        0.11,
        0.15,
    ]
).T

kalman_state, P = kalman_filter(
    kalman_state, P, F, H, R, Q, measurement[:-1], measurement[-1]
)

print(kalman_state)
